### PR TITLE
Update to `codecov-action@v4`

### PR DESCRIPTION
**Pull request by CoolCat467**
This pull request updates the code coverage action to v4

Reading the [release details](https://github.com/codecov/codecov-action/releases/tag/v4.0.0), it looks like they update node runtime thing, drop support for not telling them the token to use (trio tells them the token so that's fine), and "various argument changes"

### CODE REVIEW

**Review by mikenerone — DISMISSED** [review]

**Comment by A5rocks** [comment]
Surprised at the "-3%" coverage -- is there some platform where things broke?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -197,7 +197,7 @@ jobs:
       - name: Run tests
         run: source .venv/bin/activate && ./ci.sh
       - if: always()
-        uses: codecov/codecov-action@v3
+        uses: codecov/codecov-action@v4
```
[the remark this answers is not in the document]
Comment (CoolCat467) [reply .github/workflows/ci.yml]: Changed noted line in 4ab573e

**Comment by A5rocks** [comment]
Is codecov just consistently reporting less coverage? It looks like -3% coverage on all platforms.

**Comment by A5rocks** [comment]
I tried a workaround and it didn't work. This looks like a known issue with v4 action:
 - https://github.com/codecov/codecov-action/issues/1287
 - https://github.com/codecov/engineering-team/issues/1143 (possibly)

Consider me displeased that somehow codecov managed to have 2 bugs that effect us in a single major version bump >:[

**Comment by jakkdl** [comment]
the alpine v3 codecov action is also saying
> [2024-02-13T07:23:32.835Z] ['info'] Error fetching git root. Defaulting to /__w/trio/trio/empty. Please try using the -R flag. Error: Error running external program: Error: spawnSync git ENOENT

https://github.com/python-trio/trio/actions/runs/7883042498/job/21509252192?pr=2951#step:7:25

though that's not new (v3 on non-alpine never said it), but maybe also an indication of directory trouble?

> does exclude_lines not effect the coverage.xml file but rather how codecov interprets it?

I think this is the case, since it's under the `tool.coverage.report` heading - whereas `tool.coverage.run` is for the xml. (some local testing suggests the same, modifying the report section after having generated coverage data changes the output of `coverage report`).

So this is about codecov not being able to find the `pyproject.toml` file, maybe it's because of us messing around with `empty/` and stuff?

**Comment by jakkdl** [comment]
Random idea: I realized we don't have a `codecov.yaml`, which is the codecov-specific configuration file. You don't need one, but maybe it would help resolve our problems https://docs.codecov.com/docs/codecov-yaml

**Comment by jakkdl** [comment]
@CoolCat467 updating the branch with a merge commit will ping everybody subscribed to a PR (which might be a lot, since joining the trio/ org automatically makes you watch the repository for all activity). AFAIK there's no way to disable merge-commit-notifications in particular, so even though it sometimes is useful it'd be lovely if you refrained from repeatedly bumping PRs unless there's relevant changes or something <3

**Comment by CoolCat467** [comment]
> @CoolCat467 updating the branch with a merge commit will ping everybody subscribed to a PR (which might be a lot, since joining the trio/ org automatically makes you watch the repository for all activity). AFAIK there's no way to disable merge-commit-notifications in particular, so even though it sometimes is useful it'd be lovely if you refrained from repeatedly bumping PRs unless there's relevant changes or something <3

Sorry, I just want to make sure things stay up to date, but yes I can do that.

**Comment by CoolCat467** [comment]
Ok last update was 3 months ago, and while there are not merge conflicts I think that it's reasonable to update it again now that this much time has passed.

**Comment by CoolCat467** [comment]
It's been a long time, what exactly was the issue for not using v4 again? Loss compared to previous system or something?

**Comment by A5rocks** [comment]
Yeah, the issue was decreased coverage despite the coverage files and coverage config not actually changing.

**Comment by jakkdl** [comment]
https://github.com/codecov/codecov-action/issues/1279 has been resolved, lets see if we can get this working or there are other issues blocking us

**Comment by jakkdl** [comment]
Current status:

* merging main showed -7% coverage, which was due to codecov only displaying the v3 run. It appears we cannot mix v3 and v4?
* After changing alpine to use v4 I'm now not seeing any codecov in CI
  * do we need to open a new PR for it to properly check v4?
  * Alpine is crashing again, it appears https://github.com/codecov/codecov-action/issues/1279 has not in fact been resolved for us?
* https://github.com/codecov/codecov-action/issues/1326 is still unresolved, though it's possible the behavior has changed since it was opened >5 months ago.

**Comment by jakkdl** [comment]
I think we can close this, and open a new PR trying to update to v5

**Pull request closed without merging**